### PR TITLE
Add new lint `manual_duration_calcs`

### LINKED ISSUE #6068: Manual Duration::{as_nanos,as_milis,as_secs_f64,as_secs_f32} implementation
### What it does

Recognize cases of manual re-implementation of the builtin `as_nanos` and `as_secs_f64` and `as_secs_f32` functions on `Duration`.

### Categories 

- Kind: complexity, sometimes also correctness

### Drawbacks

MSRV issues, implementing the suggestion might increase the MSRV.

### Example

```rust
let nanos = diff.as_secs() * 1_000_000_000 + diff.subsec_nanos() as u64;
let milis = diff.as_secs() * 1_000 + diff.subsec_milis() as u64;
let secs_f64 = (diff.as_secs() as f64 * 1_000.0 + diff.subsec_milis() as f64) / 100.0;
let secs_f64 = diff.as_secs() as f64 + diff.subsec_milis() as f64 / 1_000.0;
let secs_f64 = diff.as_secs() as f64 + diff.subsec_nanos() as f64 / 1_000_000_000.0;
```

Could be written as:

```rust
let nanos = diff.as_nanos() as u64;
let milis = diff.as_millis() as u64;
let secs_f64 = diff.as_secs_f64();
let secs_f64 = diff.as_secs_f64();
let secs_f64 = diff.as_secs_f64();
```

### PR DESCRIPTION
**Pull request by himanoa**
fixes #6068
 
- \[x] Followed [lint naming conventions][lint_naming]
- \[x] Added passing UI tests (including committed `.stderr` file)
- \[x] `cargo test` passes locally
- \[x] Executed `cargo dev update_lints`
- \[x] Added lint documentation
- \[x] Run `cargo dev fmt`

[lint_naming]: https://rust-lang.github.io/rfcs/0344-conventions-galore.html#lints

Note that you can skip the above if you are just opening a WIP PR in
order to get feedback.

Delete this line and everything above before opening your PR.

---

changelog: Add new lint `manual_duration_calcs`

### CODE REVIEW

**Comment by giraffate** [comment]
It seems to fail dogfood test.

**Comment by himanoa** [comment]
I'm doing refactoring now to do pass the dogfood test.

File: clippy_lints/src/manual_duration_calcs.rs [hunk 1]
```diff
@@ -0,0 +1,566 @@
...
+            Spanned {
+                node: BinOpKind::Add, ..
+            },
+            ref left,
+            ref right,
+        ) = expr.kind
+        {
+            if_chain! {
+                if let Some((mul_ident, mul_receiver, multiplier, plus_ident, plus_receiver, plus_cast_type)) =
+                    match (&left.kind, &right.kind) {
```
Comment (flip1995) [inline clippy_lints/src/manual_duration_calcs.rs:138]: Please refactor this out in an extra function.

File: clippy_lints/src/manual_duration_calcs.rs [hunk 2]
```diff
@@ -0,0 +1,566 @@
...
+                            suggested_fn
+                        ),
+                        applicability,
+                    );
+                }
+            }
+        }
+
+        #[derive(Debug)]
+        enum Number {
```
Comment (flip1995) [inline clippy_lints/src/manual_duration_calcs.rs:472]: Also this. enums in fn in impls is a bit too deep IMO.

File: clippy_lints/src/lib.rs [hunk 3]
```diff
@@ -1107,7 +1107,7 @@ pub fn register_plugins(store: &mut rustc_lint::LintStore, sess: &Session, conf:
     store.register_late_pass(|| box inherent_impl::MultipleInherentImpl::default());
     store.register_late_pass(|| box neg_cmp_op_on_partial_ord::NoNegCompOpForPartialOrd);
     store.register_late_pass(|| box unwrap::Unwrap);
-    store.register_late_pass(|| box duration_subsec::DurationSubsec);
+    store.register_late_pass(|| box manual_duration_calcs::ManualDurationCalcs);
```
Comment (flip1995) [inline clippy_lints/src/lib.rs:1110]: You also have to add a `register_renamed` here.

**Review by flip1995 — CHANGES_REQUESTED** [review]
Yes, please do some refactoring. Splitting things up in functions will help a lot.

File: clippy_lints/src/lib.rs (hunk 3 above)
Comment (himanoa) [reply clippy_lints/src/lib.rs:1110]: Ok. I'll fix it

File: tests/ui/manual_duration_calcs.stderr [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+error: calling `as_secs_f64()` is more concise than this calculation
+  --> $DIR/manual_duration_calcs.rs:33:23
+   |
+LL |     let _secs_f64_1 = (dur.as_secs() as f64 * 1_000.0 + dur.subsec_millis() as f64) / 1000.0;
+   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `dur.as_secs_f64()`
+
+error: no manual re-implementationa of the as_millis
+  --> $DIR/manual_duration_calcs.rs:33:23
+   |
+LL |     let _secs_f64_1 = (dur.as_secs() as f64 * 1_000.0 + dur.subsec_millis() as f64) / 1000.0;
```
Comment (giraffate) [inline tests/ui/manual_duration_calcs.stderr:54]: It seems to fail tests at this line and next line. I haven't checked that in detail, but the lint is emitted twice here and rustfix seems to fail.

Comment (himanoa) [reply tests/ui/manual_duration_calcs.stderr:54]: I know rustfix conflicting between `manual_re_implementation_lower_the_unit` and `duration_as_secs` when this pattern.

```rust
let _secs_f64_1 = (dur.as_secs() as f64 * 1_000.0 + dur.subsec_millis() as f64) / 1000.0
```

I am in trouble at can't resolve this conflict.

Comment (giraffate) [reply tests/ui/manual_duration_calcs.stderr:54]: IMHO, how about checking if parent expression in `manual_re_implementation_lower_the_unit` sees `duration_as_secs`?

**Comment by flip1995** [comment]
Those are a huge amount of changes to the code, introducing multiple functions. It would take me a day to understand what every function/struct does (i.e. what pattern it matches). Please add more (doc) comments explaining the functions and structs introduced in this PR.

**Comment by giraffate** [comment]
ping from triage @himanoa. Can you add these?
https://github.com/rust-lang/rust-clippy/pull/6490#issuecomment-761837941
>Please add more (doc) comments explaining the functions and structs introduced in this PR.

**Comment by giraffate** [comment]
ping from triage @himanoa. According to the triage procedure, I'm closing this because 2 weeks have passed with no activity. If you have more time to work on this, feel free to reopen this.

**Pull request closed without merging**